docs/rfcs/035-physical-replication.md

```diff
@@ -0,0 +1,197 @@
+
+To make things worse, we initially missed this issue, and always
+started accepting queries at replica startup, even if it didn't have
+the transaction information. That could lead to incorrect query
+results and data corruption later. However, as we fixed that, we
+introduced a new problem compared to what we had before: previously
+the replica would always start up, but after fixing that bug, it migh
+tnot. In a superficious way, the old behavior was better (but could
+lead to serious issues later!). That made fixing that bug was very
+hard, because as we fixed it, we made things (superficially) worse for
+others.
+
+See https://github.com/neondatabase/neon/pull/7288 which fixed the
+bug, and follow-up PRs https://github.com/neondatabase/neon/pull/8323
+and https://github.com/neondatabase/neon/pull/8484 to try to claw back
+the cases that started to cause trouble as fixing it. As of this
+writing, there are still cases where a replica might not immediately
+start up, causing the control plane operation to fail.
+
+One long-term fix for this is to switch to using so-called CSN
+snapshots in read replica. That would make it unnecessarily to have
+the full in-progress transaction list in the replica at startup
+time. See https://commitfest.postgresql.org/48/4912/ for a
+work-in-progress patch to upstream to implement that.
+
+### Recovery conflicts and Hot standby feedback
+
+It's possible that a tuple version is vacuumed away in the primary,
+even though it is still needed by a running transactions in the
+replica. This is called a "recovery conflict", and PostgreSQL provides
+various options for dealing with it. By default, the WAL replay will
+wait up to 30 s for the conflicting query finish. After that, it will
+kill the running query, so that the WAL replay can proceed.
+
+Another way to avoid the situation is to enable `hot_standby_feedback`
+option. When it is enabled, the primary will refrain from vacuuming
+tuples that are still needed in the primary. That means potentially
+bloating the primary, which violates the usual rule that read replicas
+don't affect the operations on the primary, which is why it's off by
+default.
+
+Neon supports `hot_standby_feedback` by passing the feedback messages
+from the replica to the safekeepers, and from safekeepers to the
+primary.
+
+### Interaction with Pageserver GC
+
+The read replica can lag behind the primary. If there are recovery
+conflicts or the replica cannot keep up for some reason, the lag can
+in principle grow indefinitely. The replica will issue all GetPage
+requests to the pageservers at the current replay LSN, and needs to
+see the old page versions.
+
+If the retention period in the pageserver is set to be small, it may
+have already garbage collected away the old page versions. That will
+cause read errors in the compute, and can mean that the replica cannot
+make progress with the replication anymore.
+
+This is an unsolved problem at the moment. A "lease" mechanism is in
```

Comment:
> This is an unsolved problem at the moment

How does this relate to `standby_horizon` in the pageserver code?

```
    Propagate standby apply LSN to pageserver to hold off GC.
    
    To avoid pageserver gc'ing data needed by standby, propagate standby apply LSN
    through standby -> safekeeper -> broker -> pageserver flow and hold off GC for
    it. Iteration of GC resets the value to remove the horizon when standby goes
    away -- pushes are assumed to happen at least once between gc iterations. As a
    safety guard max allowed lag compared to normal GC horizon is hardcoded as 10GB.
    Add test for the feature.
    
    Co-authored-by: Konstantin Knizhnik <knizhnik@neon.tech>
```

Reply:
> A "lease" mechanism is in the works

From what I see in the https://github.com/neondatabase/neon/pull/7994, the current plan is to only use leases for static computes. I'm not sure we want to use leases for lagging following replicas, do we? I think it can only hide other problems, and lagging RO is not what people usually expect from the replica. IMO, force-restart if it lagged too far behind is better

> safety guard max allowed lag compared to normal GC horizon is hardcoded as 10GB

I recall that we've seen a couple of RO endpoints lagged for >10 GB, and they were trying to get GCed pages. Example https://neondb.slack.com/archives/C04DGM6SMTM/p1720801650143579

It's not enough to say that mentioned mechanism works properly, probably, but there are non-covered cases for sure

Thanks for referencing this! I didn't know about it. It'd be a good idea to add a link to the doc

Reply:
> How does this relate to standby_horizon in the pageserver code?

Oh cool, I didn't realize we did that. I'll add it to the doc.